feat(context): refactor passing navigation context

thymikee opened the pull request:
**Summary**

This is a refactor proposal for passing `navigation` through context, using the new API through `create-react-context`. 
The change is motivated by Header being rendered outside SceneView (which was a context provider for `navigation`), not getting the `navigation` prop when it was rendered as a root view, even though wrapped with `withNavigation` HOC.

Fixes #3524

**Test plan**

I'll yet need to come up with proper testing for this.

cc @satya164 as it was his idea for the context overhaul and I'm sure he has more background on the topic.

brentvatne commented:
hi @thymikee! can you rebase against master? I pushed a change to add `isFocused` to master (had it on 1.x branch already)

brentvatne commented:
waiting for review from @satya164

satya164 commented:
I'll review today

thymikee commented:
@brentvatne what do you think, should we go in direction proposed by @satya164?

brentvatne commented:
yeah that sounds good to me

brentvatne commented:
@thymikee - ping :) would love to get this in for 2.0

thymikee commented:
I've rebased and added a test. I think we can implement what @satya164 said as a followup, as I'm not 100% sure how to approach this and if I actually understand the problem correctly.

satya164 commented on src/views/withNavigationFocus.js:
```diff
@@ -36,25 +39,17 @@ export default function withNavigationFocus(Component) {
       return (
         <Component
           {...this.props}
+          navigation={this.props.navigation}
```
This should not be needed

satya164 commented on src/views/Drawer/DrawerScreen.js:
```diff
@@ -12,11 +13,9 @@ class DrawerScreen extends React.PureComponent {
-        component={Content}
-        navigation={descriptor.navigation}
-      />
+      <NavigationProvider value={descriptor.navigation}>
```
Will be nice to move this wrapper to `SceneView` to avoid repetition

satya164 commented on src/views/withNavigationFocus.js:
```diff
@@ -2,26 +2,29 @@ import React from 'react';
-      navigation: propTypes.object.isRequired,
-    };
-
     constructor(props, context) {
```
There shouldn't be a `context` argument anymore.

satya164 commented:
Also need to address:

> For header and tab, we probably want to wrap individual items such as headerLeft, headerRight, tabBarIcon etc. with the HOC to provide the same navigation prop as the screen.

thymikee commented:
>For header and tab, we probably want to wrap individual items such as headerLeft, headerRight, tabBarIcon etc. with the HOC to provide the same navigation prop as the screen.

Do you say they should be wrapped with `withNavigation` by default, so one don't have to do it anymore?

satya164 commented:
> Do you say they should be wrapped with withNavigation by default, so one don't have to do it anymore?

With NavigationProvider so correct navigation is available

satya164 commented:
@brentvatne let's merge it now since it's an improvement over previous code anyway (support for React 16.3.1 due to new context API). we can address other stuff later.
